# APK 6725 — AI for Sport & Movement Sciences
## Assignment 4: NumPy Arrays and Introduction to Pandas

This assignment assesses your ability to work with **NumPy arrays** and to run a first analysis with **Pandas**. You start from one raw force recording, then move to a table of 306 subjects. Write your code in the cells provided and test it.

Total: 5 points.

## Run this first

Run the cell below, at the start of every session. It downloads the course files to your Google Drive and keeps them up to date. When you see "Setup complete", you are ready to work.

**Do not edit or save your work inside the `AIforMovementScience` folder.** Use **File > Save a copy in Drive** and work on your own copy.

In [1]:
# [course-setup] Run this cell first, every session (or use Runtime > Run all)
import os, sys
os.chdir('/content')                      # start from a valid directory
from google.colab import drive
drive.mount('/content/drive')

REPO = '/content/drive/MyDrive/AIforMovementScience'
if os.path.isdir(REPO + '/.git'):
    # already downloaded before: just update it
    !git -C "$REPO" fetch origin main && git -C "$REPO" reset --hard origin/main
else:
    # first time, or a broken/partial copy: download a fresh copy
    !rm -rf "$REPO"
    !git clone https://github.com/dguari1/AIforMovementScience.git "$REPO"

WORKDIR = REPO + '/Assignments'          # this notebook's folder in the course files
os.chdir(WORKDIR)
sys.path.insert(0, WORKDIR)           # so local .py files (e.g. FunctionExamples.py) import correctly
DATA = REPO + '/lectures/data'        # the folder that holds this assignment's data files
print('Setup complete. You can now run the rest of the notebook.')

Mounted at /content/drive
From https://github.com/dguari1/AIforMovementScience
 * branch            main       -> FETCH_HEAD
error: cannot lock ref 'refs/remotes/origin/main': Unable to create '/content/drive/MyDrive/AIforMovementScience/.git/refs/remotes/origin/main.lock': File exists.

Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.
 ! ddc8128..fb3adea  main       -> origin/main  (unable to update local ref)
Setup complete. You can now run the rest of the notebook.


## The data

All three files come from the Gait in Parkinson's Disease database (PhysioNet, <https://physionet.org/content/gaitpdb/>). They are already in the course files, in `lectures/data`. The setup cell stored that path in the variable `DATA`.


The files **`SiCo01_01.txt` and `SiPt02_01.txt`** are two raw force recordings.

Each subject walked for about two minutes using an insole with eight force sensors under each foot. The sensors measure the vertical force between the foot and the floor, in Newtons. The signal is sampled at 100 Hz. Each file is a plain table of numbers with 19 columns:

| Column | Contents |
| --- | --- |
| 0 | time, in seconds |
| 1 to 8 | the eight sensors under the **left** foot |
| 9 to 16 | the eight sensors under the **right** foot |
| 17 | total force under the left foot |
| 18 | total force under the right foot |

`SiCo01` is a control subject: 53 years old, 56 kg.
`SiPt02` has Parkinson's disease: 77 years old, 67 kg.


The file **`gait_measures_PD_controls_ALL.csv`** contains summary data for 306 subjects, there is one subject per row.

The file contains eighteen gait measures already computed from recordings like the two above: walking speed, cadence (steps per minute), step length, stride and stance times, and the coefficient of variation (`cv_`) of each timing measure. The `class` column is `0` for a control subject and `1` for a subject with Parkinson's disease. The `ID` column names the subject who produced the recording.

The file **`demographics_Parkinsons.csv`** includes age, height, weight and clinical scores for the 64 subjects whose `ID` starts with `Si`.

## How to submit

Before you submit, please make sure your notebook is organized and reproducible. Part of your grade is for this.

- Answer each question in its own section, using the headings already provided.
- Put your code in the code cells; when a question asks for a written answer, use the markdown answer cell.
- Delete any scratch or experimental cells that are not part of your answer.
- Download your notebook: **File > Download > Download .ipynb**, name it `YourName_Assignment_4.ipynb`, and upload it to Canvas.

## Question 1 (1 point)

Run the cell below. It reads `SiCo01_01.txt` into an array called `force`. Then answer parts (a) to (c), each in its own cell.

`np.loadtxt` reads a plain table of numbers straight into an array. It is new to you; everything you do with the array afterwards is not.

In [2]:
import numpy as np
np.set_printoptions(legacy='1.21')   # optional, it makes the output easier to read

force = np.loadtxt(f"{DATA}/SiCo01_01.txt")
print('Loaded', force.shape[0], 'samples')

Loaded 12119 samples


### Part (a) — Inspect the array

Print the following characteristic of the array:
- number of dimensions,
- the shape,
- the number of elements,
- data type of `force`.

The first column of the array contains the time in seconds, print the duration of the recordings (time of the last sample - time of the last sample)


In [ ]:
# Characteristics of the array
print('number of dimensions:', force.ndim)
print('shape:', force.shape)
print('number of elements:', force.size)
print('data type:', force.dtype)

# Duration of the recording: time of the last sample - time of the first sample
duration = force[-1, 0] - force[0, 0]
print(f'duration of the recording: {duration:.2f} s')

**Your answer:** `force` is a 2-dimensional array with shape (12119, 19): 12119 samples (rows) and 19 columns. It holds 12119 × 19 = 230,261 elements of type `float64`. The recording lasts 121.17 s (about two minutes), which matches the description.

### Part (b) — Split the array

Using **slicing only, no loops**, create five arrays from `force`:

| Name | Contents |
| --- | --- |
| `time` | the time column, one-dimensional |
| `left` | the eight left-foot sensors |
| `right` | the eight right-foot sensors |
| `total_left` | the total left force, one-dimensional |
| `total_right` | the total right force, one-dimensional |


For instace, to extract the first column of the array, you can use
```Python
time = force[:,0]
```

Use slices to extract multiple columns at the same time.

Print the shape of each array. Check that the number of columns you pulled out adds up to 19.

In [ ]:
time        = force[:, 0]       # column 0
left        = force[:, 1:9]     # columns 1 to 8
right       = force[:, 9:17]    # columns 9 to 16
total_left  = force[:, 17]      # column 17
total_right = force[:, 18]      # column 18

print('time:       ', time.shape)
print('left:       ', left.shape)
print('right:      ', right.shape)
print('total_left: ', total_left.shape)
print('total_right:', total_right.shape)

# 1-D arrays count as one column each
n_columns = 1 + left.shape[1] + right.shape[1] + 1 + 1
print('columns pulled out:', n_columns)

### Part (c) — Confirm the sampling rate

The description says the signal is sampled at 100 Hz. Check it. Use slicing to get the time between consecutive samples, then report the mean sampling interval in seconds and the sampling frequency in Hz (1/sample_in_seconds).

**Hint:** `time[1:]` is every sample except the first, and `time[:-1]` is every sample except the last. Subtracting one from the other gives you all the gaps between consecutive sampling times.

In [ ]:
dt = time[1:] - time[:-1]      # time between consecutive samples
mean_dt = dt.mean()
fs = 1 / mean_dt

print(f'mean sampling interval: {mean_dt:.6f} s')
print(f'sampling frequency: {fs:.2f} Hz')

**Your answer:** The mean sampling interval is 0.0100 s (0.009999 s), so the sampling frequency is about 100.01 Hz. This confirms that the signal is sampled at 100 Hz.

## Question 2 (1.5 points)

Summarize the recording you loaded in Question 1. Answer parts (a) to (d), each in its own cell.

### Part (a) — Load carried by each sensor

For the eight left-foot sensors in `left`, compute the mean of each sensor with **one** call, and the maximum of each sensor with **one** call. Use the `axis` argument; do not write a loop.

In the answer cell, say which of the eight left sensors carries the most load on average.

In [ ]:
left_mean = left.mean(axis=0)   # axis=0 collapses the rows: one mean per sensor (column)
left_max  = left.max(axis=0)    # one maximum per sensor

print('mean of each left sensor (N):', left_mean)
print('max of each left sensor (N): ', left_max)

best = np.argmax(left_mean)
print(f'sensor with the highest mean load: sensor {best + 1} (index {best} in left), {left_mean[best]:.1f} N')

**Your answer:** Left sensor 7 (index 6 in `left`, column 7 of `force`) carries the most load on average, about 90.6 N. It also has the highest peak (501.3 N). Sensor 1 is second at about 79.6 N on average.

### Part (b) — Express force as a percentage of body weight

Force in Newtons depends on how heavy the subject is. To compare subjects you first divide by body weight.

`SiCo01` has a mass of 56 kg. Convert that to a weight in Newtons (mass times 9.81), then use **broadcasting** to turn `total_left` and `total_right` into percentages of body weight. Report the peak value of each foot.

In [ ]:
mass_kg = 56.0
body_weight_N = mass_kg * 9.81
print(f'body weight: {body_weight_N:.1f} N')

# rememebrt that one number divides a whole array: that is broadcasting

# Write your code here
total_left_bw  = total_left  / body_weight_N * 100
total_right_bw = total_right / body_weight_N * 100

print(f'peak left foot:  {total_left_bw.max():.1f} % of body weight')
print(f'peak right foot: {total_right_bw.max():.1f} % of body weight')

### Part (c) — Find when each foot is on the floor

A foot is on the floor when the total force under it is above 20 N.

Use the boolean mask provided for each foot (`loaded_left` and `loaded_right`) and report the fraction of the recording in which each foot is loaded. **Hint:** the mean of a boolean array is the fraction of `True` values.

In the answer cell: the two fractions add up to more than 1. Explain how that is possible.

In [ ]:
loaded_left  = total_left  > 20
loaded_right = total_right > 20
#note that you need to define the variable total_left and total_right before using them in the above code.

# Write your code here
frac_left  = loaded_left.mean()
frac_right = loaded_right.mean()
print(f'fraction of the recording with the left foot loaded:  {frac_left:.3f}')
print(f'fraction of the recording with the right foot loaded: {frac_right:.3f}')
print(f'sum of the two fractions: {frac_left + frac_right:.3f}')

# time when both feet are on the floor at the same time (double support)
both = loaded_left & loaded_right
print(f'fraction with both feet loaded: {both.mean():.3f}')

**Your answer:** The left foot is loaded 0.659 of the time and the right foot 0.651, which adds up to 1.31. That is possible because the fractions overlap: in walking there are periods of **double support**, when both feet are on the floor at the same time (at the transition from one step to the next). Both feet are loaded about 0.31 of the recording, and that time is counted in both fractions. Only a gait with no double support, such as running, would give a sum of 1 or less.

### Part (d) — Average the force second by second

The signal is sampled at 100 Hz, so every block of 100 samples is one second of walking.

The following code trims `total_left_bw` so that its length is a whole multiple of 100. It then reshapes it so that each row holds one second of data.

Using this result, compute the mean force in each second. Report the shape of the result, its mean and its standard deviation


In [ ]:
n = (len(total_left_bw) // 100) * 100
reshaped_left  = total_left_bw[:n].reshape(-1, 100)

# Write your code here
per_second_left = reshaped_left.mean(axis=1)   # axis=1: the mean of each row, i.e. of each second

print('shape of the result:', per_second_left.shape)
print(f'mean: {per_second_left.mean():.2f} % of body weight')
print(f'standard deviation: {per_second_left.std():.2f} % of body weight')

## Question 3 (0.5 points)

`SiPt02_01.txt` is a second recording, from a subject with Parkinson's disease, mass 67 kg. The file has exactly the same 19 columns. Answer parts (a) and (b), each in its own cell.

### Part (a) — Wrap your work in a function

The following function takes a file name and a body mass, and returns:

- the peak force under each foot, in percentage of body weight
- the fraction of the recording in which each foot is loaded
- the mean and the standard deviation of the per-second force under the left foot, in percentage of body weight

Execute the following cell to ensure the function exists in the notebook memory

In [ ]:
def summarize(file_name, mass_kg):
    """Summarize one force recording. Returns a dictionary."""
    data = np.loadtxt(file_name)
    tot_left, tot_right = data[:, 17], data[:, 18]

    body_weight_N = mass_kg * 9.81
    left_bw  = tot_left  / body_weight_N * 100
    right_bw = tot_right / body_weight_N * 100

    n = (len(left_bw) // 100) * 100
    per_second = left_bw[:n].reshape(-1, 100).mean(axis=1)

    return {
        'peak_left_bw'   : left_bw.max(),
        'peak_right_bw'  : right_bw.max(),
        'loaded_left'    : (tot_left  > 20).mean(),
        'loaded_right'   : (tot_right > 20).mean(),
        'per_second_mean': per_second.mean(),
        'per_second_std' : per_second.std(),
        'per_second'     : per_second,
    }

Read the function and get a complete understanding of what it is doing.

Call the function `summarize` on both subjects and print the two sets of results.

The function returns a dictionary, use a for loop so that the results from each subject are side by side.

You results should look something like this:
```Python
measure              control        PD
peak_left_bw         XXX            XXX
peak_right_bw        XXX            XXX
loaded_left          XXX            XXX
loaded_right         XXX            XXX
per_second_mean      XXX            XXX
per_second_std       XXX            XXX
```

where `XXX` are the results obtained for each number.

In [ ]:
control = summarize(f"{DATA}/SiCo01_01.txt", 56)
pd_subj = summarize(f"{DATA}/SiPt02_01.txt", 67)

measures = ['peak_left_bw', 'peak_right_bw', 'loaded_left', 'loaded_right',
            'per_second_mean', 'per_second_std']

print(f"{'measure':<20}{'control':>10}{'PD':>10}")
for measure in measures:
    print(f"{measure:<20}{control[measure]:>10.3f}{pd_subj[measure]:>10.3f}")

In the answer cell: name one measure where the two subjects differ and one where they look alike. Explain your result.

**Your answer:**

- **Where they differ:** `per_second_std`, the variability of the per-second left-foot force. It is 14.0 % of body weight for the control subject and 7.8 % for the PD subject, almost twice as large. The one-second windows do not match the length of a stride. The control subject's strides last about 1.27 s, so each window captures a different part of the step, and the per-second mean moves up and down a lot. The PD subject's strides last about 1.10 s, close to one second, so each window contains nearly one full stride and the per-second means are steadier. This difference reflects the walking rhythm and the choice of window. It does not mean the PD subject walks more steadily. The per-second mean is also higher for the PD subject (72.0 vs 65.7 %), in line with a slightly longer loaded time.
- **Where they look alike:** `loaded_left` and `loaded_right`. Each foot is loaded about 65 % of the time for the control subject and about 69–70 % for the PD subject, and each subject is symmetric between left and right. The peak forces are also similar once normalized to body weight (about 144 % vs 150–152 %). Dividing by body weight removes most of the difference between a 56 kg and a 67 kg person.

These are only two subjects, one per group, so these differences cannot be generalized to PD.

## Question 4 (1.5 points)

Two subjects are not evidence. `gait_measures_PD_controls_ALL.csv` holds the same kind of measures for 306 walking recordings, already computed. Move to Pandas and answer parts (a) to (c), each in its own cell.

### Part (a) — Read the table and look at it

Read `gait_measures_PD_controls_ALL.csv` into a DataFrame with `pd.read_csv`. The file is at `f"{DATA}/gait_measures_PD_controls_ALL.csv"`.

Remember to import pandas first. Print the shape, the first five rows (using `.head()`), and `.describe()` for `avg_speed`, `cadence` and `avg_step_length`. Then report how many rows are in each `class`.

In [ ]:
import pandas as pd

gait = pd.read_csv(f"{DATA}/gait_measures_PD_controls_ALL.csv")

print('shape:', gait.shape)
display(gait.head())
display(gait[['avg_speed', 'cadence', 'avg_step_length']].describe())

print('rows in each class:')
print(gait['class'].value_counts())

### Part (b) — Add two columns

Add a column `group` that holds the text `'control'` where `class` is 0 and `'PD'` where `class` is 1. Add a column `study` that holds the first two characters of `ID`.

**Hint:** use `.map({0: 'control', 1: 'PD'})` for the first one, `.str[:2]` for the second.

In [ ]:
gait['group'] = gait['class'].map({0: 'control', 1: 'PD'})
gait['study'] = gait['ID'].str[:2]

gait[['ID', 'class', 'group', 'study']].head()

### Part (c) — Compare the two groups

Group the table by `group` and report the mean of `avg_speed`, `cadence`, `avg_step_length` and `cv_stride_time_left`.

Then check this identity on the two group means:

$$ \text{speed (m/s)} = \frac{\text{cadence (steps/min)} \times \text{step length (m)}}{60} $$

confirm that this measure of `speed` correspond to the reported `avg_speed`.

In the answer cell: the two groups walk at clearly different speeds but at almost the same cadence. Cadence is steps per minute. Using the identity above, what does that tell you about **how** the slower group is slower?

In [ ]:
group_means = gait.groupby('group')[['avg_speed', 'cadence', 'avg_step_length', 'cv_stride_time_left']].mean()
display(group_means)

# check the identity: speed = cadence * step length / 60
group_means['speed_from_identity'] = group_means['cadence'] * group_means['avg_step_length'] / 60
display(group_means[['avg_speed', 'speed_from_identity']])

**Your answer:** The group means are: control 1.24 m/s, 109.2 steps/min, 0.69 m step length; PD 1.02 m/s, 108.8 steps/min, 0.57 m step length. The identity gives 109.2 × 0.688 / 60 = 1.25 m/s for controls and 108.8 × 0.565 / 60 = 1.03 m/s for PD, which matches the reported `avg_speed` (1.24 and 1.02 m/s). The small gap comes from applying the identity to group means (the mean of a product is not exactly the product of the means) and from the missing values in `avg_speed` and `avg_step_length`.

Speed = cadence × step length / 60, and cadence is almost the same in both groups (about 109 steps/min). So the slower speed of the PD group comes almost entirely from **shorter steps** (0.57 m vs 0.69 m, about 18 % shorter), not from taking fewer steps per minute. The PD group takes steps just as often but covers less ground with each one, which fits the short, shuffling steps typical of Parkinsonian gait. The PD group also has a higher `cv_stride_time_left` (0.22 vs 0.15), meaning their stride timing is more variable.

## Question 5 (0.5 points)

You now have a group difference. Before trusting it, check where it came from.

The 306 recordings come from three separate studies, marked by the `study` column you built in Question 4, part (b). The three studies used different protocols.

Build a pivot table of the mean `avg_speed`, with `study` as the rows and `group` as the columns. Also report how many rows sit in each study-and-group cell.

In the answer cell: does the speed difference appear in all three studies? Why is that worth checking?

In [ ]:
speed_table = gait.pivot_table('avg_speed', index='study', columns='group', aggfunc='mean')
display(speed_table)

# number of rows in each study-and-group cell
count_table = gait.pivot_table(index='study', columns='group', aggfunc='size')
display(count_table)

**Your answer:** Yes, the speed difference appears in all three studies. Controls walk faster than PD subjects in each one: Ga 1.23 vs 0.99 m/s, Ju 1.24 vs 1.01 m/s, Si 1.25 vs 1.11 m/s. The gap is smaller in the Si study (about 0.14 m/s vs about 0.23–0.24 m/s in the other two).

This is worth checking because the three studies used different protocols, and they do not contribute equally to each group. For example, Ju has 104 PD rows but only 25 controls. If one study's protocol made everyone walk more slowly, and that study mostly contributed PD subjects, the pooled table could show a "PD is slower" difference that really comes from the study, not the disease (a confound). Because the difference shows up within every study, it is not an artifact of how the studies were mixed, and we can trust it more.